### DISCUSSION TITLE
❓ How to pass on Selected Rows from table in Modal to parent Form ❓

### DISCUSSION THREAD

**Discussion by sv-hmelevsky** [filamentphp/filament#14655] [Help] [answered]
### Package

Form builder

### Package Version

v3.2.0

### How can we help you?
Hi everybody!
I need help! 🙏 How to pass on selected rows from table in Modal to parent Form without Bulk Action?
Like this:
![select_rows_in_modal_to_form](https://github.com/user-attachments/assets/b706ca5e-dc09-4b1c-8943-aca35fed8acd)

Modal Action:
```
TextInput::make('shelves_in_operation')
        ->label('Shelves in Operation')
        ->default(0)
        ->suffixAction(
            Action::make('selectRacks')
                ->label('Select Racks')
                ->icon('heroicon-m-ellipsis-horizontal')
                ->modalWidth(MaxWidth::Full)
                ->modalContent(fn() => view('livewire.rack-selection-table'))
                ->action(function(array $data) {
                    dd("SELECTED ROWS ARRAY[] HERE");
                })
                ->stickyModalHeader()
                ->stickyModalFooter()
        )
        ->live()
        ->required(),
```

Any help and suggestion will be appreciated!

**Comment by sprtk-ches** [comment]
https://filamentphp.com/docs/3.x/tables/actions#accessing-the-selected-table-rows

Then you might be able to use that to pass the data back (with an event or something)

**Comment by sv-hmelevsky** [comment, marked as answer]
👋 Hello everyone! 

This was a real challenge. Because the developers don't have such functionality at all. Which is very strange, loading functional tables in modal windows, with subsequent data retrieval, is a common thing in large projects! The developers should think about adding something like this in a future or current version of Filament. 🤔

In the end, the problem was solved, and I am very, very happy about it! 😊 But I'm not sure that I got an elegant solution. It looks more like a crutch to me. I had to give up buttons on the modal window, there is no way to get data from the table through them! 😔

I added a button for Bulk Action to the table. I was able to get an array of IDs of selected rows without any problems. But how do I pass it back? I tried different options, none of them worked. In the end, I settled on caching. I write the value to the cache along with the user ID, so that there are no collisions in case of its receipt. 😏

Okay, the data has been transferred, how can I now close the modal window immediately after confirming the selection? Here's a new quest! 😞

I tried to assign an ID to the modal window via `extraModalWindowAttributes()`, and after the BulkAction is triggered,
call `$livewire->dispatch('close-modal', 'myModalNameId');` But this approach does not work! I tried different options, none of them work! To close the modal window, `$livewire->dispatch('close-modal')` is called, that's a fact, but a random ID is passed there, which, as luck would have it, is **protected**! 😖 This is just a mockery! I had to build another bicycle and resort to reflection. As a result, I extracted the Id and passed it to the component. Thus, I managed to close the modal window.

Then another test, the action method, which usually works when you click the "Send" button in the modal window, is simply not called, because the modal window did not close as usual. And no matter what I did, the action was not called. Another dose of frustration. 😡

In the end, on the resource creation page: `class CreateFertilizerExpense extends CreateRecord`. I added a Livewire event, which I call immediately before closing the modal window. In the method itself, I managed to pull data from the cache and assign it to the fields. Hooray! Victory! 🌞🎉

It was quite an adventure, I tell you, I hope the developers will do something about it.
If you have any ideas on how to simplify what I did, welcome to the discussions!
I hope my "guide" will help someone solve a similar problem.

Peace! ✌️

This is the code:
Form in Ressource **FertilizerExpenseResource**:

```
TextInput::make('shelves_in_operation')
      ->label('Shelves in operation')
      ->default(0)
      ->suffixAction(
          Action::make('selectRacks')
              ->label('Select Racks')
              ->icon('heroicon-m-ellipsis-horizontal')
              ->modalWidth(MaxWidth::Full)
              ->modalContent(function($livewire) {
                  // Reflection to get ID from a private field. There is no other way, unfortunately.
                  $reflectionClass = new \ReflectionClass($livewire::class);
                  $reflectionProperty = $reflectionClass->getProperty('__id');
                  $reflectionProperty->setAccessible(true);
                  $parentModalId = $reflectionProperty->getValue($livewire);

                  return view('livewire.rack-selection-table', ['parentModalId' => $parentModalId]);
              })
              ->stickyModalHeader()
              ->modalSubmitAction(false)
              ->modalCancelAction(false)
      )
      ->live()
      ->required(),
```

Blade template **rack-selection-table.blade.php**:
`@livewire('shelves-in-operation-table', ['parentModalId' => $parentModalId])`


Livewire Component **ShelvesInOperationTable**:

```
...
class ShelvesInOperationTable extends Component implements HasForms, HasTable
{
    use InteractsWithTable, InteractsWithForms;

    public function mount($parentModalId)
    {
        $this->parentModalId = $parentModalId;
    }

    public function table(Table $table): Table
    {
        return $table
            ->query(Rack::query())
            ->columns([
                ... columns ...
            ])
            ->selectable()
            ->paginated(false)
            ->bulkActions([
                BulkAction::make('select_records')
                    ->label('Select records')
                    ->action(function(array $data, $livewire) {
                        $records = $livewire->selectedTableRecords;
                        $count = count($records);
                        
                        $shelfCountSum = Rack::whereIn('id', $records)->sum('shelf_count');
                        $user_slug = "user_id_" . Filament::auth()->user()->getAuthIdentifier();

                        Cache::set("{$user_slug}_shelves_in_operation_count", $count);
                        Cache::set("{$user_slug}_shelves_count_sum", $shelfCountSum);

                        //Name of your event
                        $livewire->dispatch('refreshShelvesInOperation');
                        $livewire->dispatch('close-modal', id: "{$this->parentModalId}-form-component-action");
                    })
            ]);
    }

    public function render(): View
    {
        return view('livewire.shelves-in-operation-table');
    }
}
```

Resource create page **CreateFertilizerExpense.php**:

```
use Livewire\Attributes\On; // <-- don't forget it

class CreateFertilizerExpense extends CreateRecord
{
    protected static string $resource = FertilizerExpenseResource::class;

    protected function getRedirectUrl(): string
    {
        return $this->getResource()::getUrl('index');
    }

    #[On('refreshShelvesInOperation')]
    public function refresh(): void
    {
        $user_slug = "user_id_" . Filament::auth()->user()->getAuthIdentifier();
        $this->data['shelves_in_operation'] = Cache::pull("{$user_slug}_shelves_in_operation_count");
        $this->data['active_shelves'] = Cache::pull("{$user_slug}_shelves_count_sum");
    }
}
```

**Comment by elmudometal** [comment]
Hi, I understand your frustration. I went through the same thing as you and I'm creating a component for that: [https://github.com/elmudometal/filament-select-table](url)
It's in a beta version, I have 2 things that I still couldn't select for the multiple view
1- being able to check the checkboxes by default exists in $selectedRecord and $selectedTableRecord but it seems they can't be set when starting the table
2- being able to send the table schema, it receives it well but livewire can't reassign it and it gets lost

If you want to try the package and give me your feedback I would appreciate it.

**Reply by sv-hmelevsky** [reply]
Hi, maybe you have a video demo of how your component works? It would be interesting to have a look at it!

**Reply by elmudometal** [reply]
I've uploaded an update to my plugin already with some bug fixes and a video of its use.

**Comment by finity-labs** [comment]
This flow is now available as a package: https://github.com/finity-labs/fin-modal-table-select

`fillsRepeater()` syncs a modal table selection into a sibling Repeater — one editable row per record — and re-confirming the modal merges instead of resetting: newly picked records are appended, deselected ones removed, and values the user already edited (quantity, price) are kept. The paired `SelectedItemsRepeater` handles the reverse direction: deleting a row deselects its record. MIT, Filament v4 and v5.
